daemon/cmd/ciliumendpoints.go

```diff
@@ -0,0 +1,116 @@
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/endpoint"
+	cilium_v2a1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	ciliumv2 "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+)
+
+type endpointCache interface {
+	LookupPodName(name string) *endpoint.Endpoint
+}
+
+// cleanStaleCEPs runs on Daemon init, attempting to delete any CiliumEndpoints that are referencing local
+// Pods that are not being managed (i.e. do not have Endpoints).
+// This must only be run after K8s Pod and CES/CEP caches are synced and local endpoint restoration is complete.
+func (d *Daemon) cleanStaleCEPs(ctx context.Context, eps endpointCache, ciliumClient ciliumv2.CiliumV2Interface, enableCiliumEndpointSlice bool) error {
+	keyFn := func(ns string, name string) string { return ns + "/" + name }
+
+	cesContainedCEPLookup := map[string]*cilium_v2a1.CoreCiliumEndpoint{}
+	if enableCiliumEndpointSlice {
+		indexer := d.k8sWatcher.GetIndexer("ciliumendpointslice")
+		// localNode index contains indices for ces referencing local endpoints.
+		cesObjs, err := indexer.ByIndex("localNode", node.GetCiliumEndpointNodeIP())
+		if err != nil {
+			return fmt.Errorf("could not get ciliumendpoints by index: %w", err)
+		}
+		for _, cesObj := range cesObjs {
+			ces, ok := cesObj.(*cilium_v2a1.CiliumEndpointSlice)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpointslice store: %T", cesObj)
+			}
+			for i := range ces.Endpoints {
+				cep := ces.Endpoints[i]
+				if cep.Networking.NodeIP == node.GetCiliumEndpointNodeIP() {
+					cesContainedCEPLookup[keyFn(ces.Namespace, cep.Name)] = &cep
+				}
+			}
+		}
+	}
+
+	pods, err := d.k8sWatcher.GetCachedPods()
+	if err != nil {
+		return fmt.Errorf("could not get cached pods: %w", err)
+	}
+
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+
+		var cepName, cepNamespace string
+		store := d.k8sWatcher.GetStore("ciliumendpoint")
+		key := keyFn(pod.Namespace, pod.Name)
+		if store == nil {
+			cep, exists := cesContainedCEPLookup[key]
+			if !exists {
+				continue
+			}
+			cepName = cep.Name
+			cepNamespace = pod.Namespace
+		} else {
+			cepObj, exists, err := store.GetByKey(key)
+			if err != nil {
+				return fmt.Errorf("could not get pod CEP from store: %w", err)
+			}
+			if !exists {
+				continue
+			}
+			cep, ok := cepObj.(*types.CiliumEndpoint)
+			if !ok {
+				return fmt.Errorf("unexpected object type returned from ciliumendpoint store")
+			}
+			cepName = cep.Name
+			cepNamespace = cep.Namespace
+		}
+		if cepName == "" || cepNamespace == "" {
+			continue
+		}
+
+		// See if local endpoint exists for this Pod.
+		podName := keyFn(cepNamespace, cepName)
```

Comment:
UUIC this `podName` is definitely the same as `key` defined on line 65. So we could simplify the previous code and not use `cepName` nor `cepNamespace`.